File: Medium/11 - Container With Most Water/11-code.cpp
#include <vector>
#include <algorithm>

/**
 * Container With Most Water
 * Sliding window approach
 * https://leetcode.com/problems/container-with-most-water/
 *
 * Created: 2024-04-04
 * Updated: 2024-04-04
 * Author: Ditra Amadia
 *
 * O(n) time complexity
 * 46ms runtime
 */

class Solution
{
public:
  int maxArea(std::vector<int> &height)
  {
    int area = 0;
    int i = 0;
    int j = height.size() - 1;

    while (j > i && j != i)
    {
      int minHeight = std::min(height[i], height[j]);
      area = minHeight > area / (j - i) ? minHeight * (j - i) : area;

      if (height[i] >= height[j])
        j--;
      else
        i++;
    }

    return area;
  }
};

File: Medium/11 - Container With Most Water/11-code.py
'''
Container With Most Water
https://leetcode.com/problems/container-with-most-water/

Created: 2025-01-31
Updated: 2025-01-31
Author: Ditra Amadia

O(n) time complexity
67ms runtime
'''

from typing import List

class Solution:
  def maxArea(self, height: List[int]) -> int:
    mx, i, j = 0, 0, len(height) - 1

    while i < j:
      area = min(height[i],height[j]) * (j - i)
      if area > mx:
        mx = area
      
      if height[i] < height[j]:
        i += 1
      else:
        j -= 1
    
    return mx

height = [1,8,6,2,5,4,8,3,7]
print(Solution().maxArea(height))